[WIP] GH-1410 Websockets ping/pong fix for proxies

Zate opened the pull request:
Added a ping from the server side to each socket (even terminal) every 30 seconds.  Client will respond with a pong.  It's not visible in the frames viewed on Chrome Developer mode but you can verify via tcpdump that ping/pong is happening.  I marked it [WIP] because it should be tested and have more eyes on it.  I think it might be doing the ping/pong multiple times per socket or not turning it off when the socket dies and adding a new one when the socket reconnects.  Ties to  #1410 

Signed-off-by: Zate Berg <zate75@gmail.com>

akosyakov commented on packages/core/src/node/messaging/connection.ts:
```diff
@@ -44,6 +44,36 @@ export function openSocket(options: IServerOptions, onOpen: OnOpen): void {
+
+            const extWs = ws as ExtWebSocket;
+
+            ws.on('pong', () => {
```
It will install a new listener each 30s, please move it in `ws.on('connection', ...` to install it once, as here; https://github.com/websockets/ws#how-to-detect-and-close-broken-connections

Zate replied:
Will do.

marcdumais-work commented:
@Zate FYI I have tested this WIP PR in our environment (that includes traefik), and it does seem to fix the issue. Thanks for this!

kittaakos commented on packages/core/src/node/messaging/connection.ts:
```diff
@@ -44,6 +44,36 @@ export function openSocket(options: IServerOptions, onOpen: OnOpen): void {
+        isAlive: boolean;
+    }
+
+    wss.on('connection', (ws: ws) => {
```
To avoid warnings due to the shadowed names of `ws`, you could change the argument to `websocket` or whatever that does not clash with `ws`. Also, since you cast the argument, you can remove the `ws` type information too. Something like this:
```typescript
wss.on('connection', websocket => {
  const extWs = websocket as ExtWebSocket;
  extWs.isAlive = true;
});
```

Same [below](https://github.com/theia-ide/theia/pull/1454/files#diff-906d0f57b69cd6a3345df7a0bda73360R59).

marcdumais-work commented:
I think the only thing pending is the comment from @akosyakov about the listener: https://github.com/theia-ide/theia/pull/1454#pullrequestreview-101873134

kittaakos commented:
@Zate, could you please update your PR based on the feedback so that we can merge your changes after a final verification?

Zate commented:
Yep working on this tonight, I've been on vacation for 5 days :)  should have something soon.

akosyakov commented:
@Zate Moin, could you just rebase your previous changes on the current master? it was good.

marcdumais-work commented:
@Zate closed by accident?

Zate commented:
Yes closed by accident.

Deleted the branch i was working on as it was cluttered with various failed attempts to resolve issues.  I will do a new pull request once I have it working.  I'm hitting some issues with the order of logic and it creating too many ping/pong checks.  I rebased a new GH-1410 branch to work from with just my changes.  Getting it to build on windows is a pain, might need to move into a linux vm.  Sorry, hope to have something working today.

Zate commented:
@akosyakov the original code was ok?  If so I can rebase and do a new pull request.

marcdumais-work commented:
@Zate no worries - thank you for persevering with this patch.

kittaakos commented:
> Getting it to build on windows is a pain

@Zate, let us know if you need assistance.

Zate commented:
It's working for the purpose of keeping the sockets open, however I am tracking down some issues with `MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 pong listeners added. Use emitter.setMaxListeners() to increase limit []` from the node console which I think might be caused by the fact that I do setInterval() and wss.clients.forEach on every "openSocket" and so I think we are setting up multiple ping/pong listeners per socket.  Using Chromes `chrome://net-internals/#events&q=type:SOCKET%20is:active` to look at the active sockets for localhost:3000 I am seeing lots of ping/pongs per socket.  I feel like perhaps setting the setInterval() might be better off happening somewhere inside the on.upgrade area, or perhaps remove the wss.clients.forEach and just set it for the socket that was created as part of openSocket().  I'm doing some trial and error to see if I can keep the functionality that works but get rid of the excessive number of ping/pongs as I think it will cause many issues over time.

Zate commented:
I was not able to get that code working correctly, I had to make a bunch of changes to remove errors and then it would kill the sockets at the time (10s I was testing with).  I have gotten a working WIP on #1534 that I think should work.

PavelSosin commented:
I suppose that it will work out-off-the box if browser will connect to LSP or DAP Servers directly as it is intended by WebSocket protocol. WebSocket connection from Chrome / EdgeChromium WebWorkers is something awesome. It sends pong automatically. WebWorker is aware about LAN disconnections via online and offline events. In opposite, You can find many WebSocket related issues in nodejs based proxies. using Google search. Two application should use other then WebSocket protocol otherwise it can break Cloud security which is protocol-aware.
